### PR TITLE
Pulse model - add PulseTarget, Port

### PR DESCRIPTION
**Pull request by TsafrirA**
### Summary
This PR completes #10694, and adds a `Port` class to allow control over specific ports. Both `Port` and `LogicalElement` are grouped under the class `PulseTarget`.

### Details and comments
Following #10694 and #10759 concerns were raised about the abstraction level of the `LogicalElement` class, and the lack of ability to control the system on a more direct level. In response, and while `LogicalElement` should still be seen as the preferred way to write pulse programs, the `Port` class was added to allow a more direct control over the hardware. This level of control is meant for users "located" closer to the hardware.

Release notes will be added at a later stage of the project.

### CODE REVIEW

File: qiskit/pulse/model/pulse_target.py [hunk 1]
```diff
@@ -20,15 +20,80 @@
...
+    """Base class of pulse target.
+
+    When playing a pulse on a quantum hardware, one typically has to define on what hardware component
+    the pulse will be played, and the frame (frequency and phase) of the carrier wave.
+    :class:`PulseTarget` addresses the first of two, and identifies the component which is the target
+    of the pulse. Every played pulse and most other instructions are associated with a
+    :class:`PulseTarget` on which they are performed.
+    """
+
+    def __init__(self, hash_identifier):
```
Comment (nkanazawa1989) [inline qiskit/pulse/model/pulse_target.py]: This enforces its subclass to define a single hashable object. Seems like this removes flexibility of class design. I think it's better to turn `__hash__` into abstract method, and let each subclass explicitly define hash within its scope. Is the precache critical for performance?

File: qiskit/pulse/model/__init__.py [hunk 2]
```diff
@@ -12,33 +12,41 @@
...
 the different combinations of :class:`.LogicalElement` and :class:`.Frame`
 to hardware aware objects is left to the Pulse Compiler.
 
-.. _logical_elements:
+.. _pulse_targets:
 
-LogicalElement
+PulseTarget
 ================
+:class:`.PulseTarget` includes :class:`.Port` who's objects are identified by a string, and
```
Comment (nkanazawa1989) [inline qiskit/pulse/model/__init__.py]:
```suggestion
:class:`.PulseTarget` includes :class:`.Port` who's objects are identified by a unique string identifier defined by the control system, and
```

File: qiskit/pulse/model/pulse_target.py [hunk 3]
```diff
@@ -20,15 +20,80 @@
...
+
+    def __hash__(self) -> int:
+        return self._hash
+
+
+class Port(PulseTarget):
+    """A ``Port`` type ``PulseTarget``.
+
+    A :class:`Port` is the most basic ``PulseTarget`` - simply a hardware port responsible for
+    pulse emission. A :class:`Port` is identified by a string, which must be recognized by the
```
Comment (nkanazawa1989) [inline qiskit/pulse/model/pulse_target.py]:
```suggestion
    A :class:`Port` is the most basic ``PulseTarget`` - simply a hardware port responsible for
    pulse emission. A :class:`Port` is identified by a string, which must be recognized by the
```
We don't have port for signal acquisition?

File: qiskit/pulse/model/pulse_target.py [hunk 4]
```diff
@@ -20,15 +20,80 @@
...
+    backend. Therefore, using pulse level control with :class:`Port` requires an extensive
+    knowledge of the hardware. When possible, it is recommended to use :class:`LogicalElement`,
```
Comment (nkanazawa1989) [inline qiskit/pulse/model/pulse_target.py]:
```suggestion
    knowledge of the hardware. When possible, it is recommended to use :class:`LogicalElement`,
```
I don't think we need to recommend some use pattern in the class documentation (this is more like tutorial). We just need to tell the identifier cannot be determined by the end users, and user-provided identifier will just fail in execution. e.g. if you are a pulse compiler pass developer, dealing with `Port` type might be more convenient.

**Review by nkanazawa1989 — COMMENTED** [review]
Thanks Tsafrir. The new class hierarchy looks reasonable, and having `Port` type is also beneficial to the pulse compiler.

File: qiskit/pulse/model/pulse_target.py (hunk 1 above)
Comment (TsafrirA) [reply qiskit/pulse/model/pulse_target.py]: Honestly, I just didn't want to have the class empty. I don't think the precache is crucial here.

File: qiskit/pulse/model/pulse_target.py (hunk 4 above)
Comment (TsafrirA) [reply qiskit/pulse/model/pulse_target.py]: Thanks.

File: qiskit/pulse/model/pulse_target.py (hunk 3 above)
Comment (TsafrirA) [reply qiskit/pulse/model/pulse_target.py]: Good point. Thanks.

**Comment by Raghav-Bell** [comment]
Linting tests are failing, please check linting tests locally. (refer: [Contribution guide](https://github.com/Qiskit/qiskit/blob/main/CONTRIBUTING.md))

**Review by nkanazawa1989 — APPROVED** [review]
LGTM. Thanks @TsafrirA

**Pull request merged**